### PR TITLE
[Proof Validation] Move proof validation to session settlement

### PR DESCRIPTION
**Pull request by Olshansk**
## Summary

⚠️  WARNING: THIS IS NOT AN EASY PR TO REVIEW ⚠️ 
- A lot of components are moving aroud
- This is sensitive since it's related to proof validation
- Best effort is all that's expected

What was changed?
- Move proof validation from `SubmitProof` in the msgServer handler to `SessionSettlement` during the EndBlocker.

Why was it changed?
1. RPC requests should be quick, lightweight and only do basic validation
2. Validators are the ones responsible for the heavy processing & validation during state transitions
3. This creates an opportunity to slash suppliers who submit false proofs, whereas
     they can keep retrying if it takes place in the SubmitProof handler.

Other:
- A few other changes related to test & test helper improvement were made
- Some improvements to the claim & proof creation logic were made

## Issue

- One of the follow ups identified while reviewing #690

## Type of change

Select one or more:

- [x] New feature, functionality or library
- [ ] Bug fix
- [x] Code health or cleanup
- [ ] Documentation
- [ ] Other (specify)

## Testing
- [ ] **Unit Tests**: `make go_develop_and_test`
- [ ] **LocalNet E2E Tests**: `make test_e2e`
- [ ] **DevNet E2E Tests**: Add the `devnet-test-e2e` label to the PR.

## Sanity Checklist

- [ ] I have tested my changes using the available tooling
- [ ] I have commented my code
- [ ] I have performed a self-review of my own code; both comments & source code
- [ ] I create and reference any new tickets, if applicable
- [ ] I have left TODOs throughout the codebase, if applicable

## Summary by CodeRabbit

- **New Features**
	- Introduced an expiration reason for claims to enhance clarity on claim status.
	- Added new error handling for non-existent accounts in proof submissions.
	- Implemented comprehensive proof validation and related tests to improve reliability.
	- Expanded interfaces for account and proof management capabilities.

- **Bug Fixes**
	- Enhanced error handling for account retrieval to prevent null-related issues.

- **Tests**
	- Improved test structure for claim settlement and proof validation.

- **Documentation**
	- Updated comments and structured the code for better readability and maintainability.

### CODE REVIEW

File: x/proof/keeper/proof_validation_test.go [hunk 1]
```diff
@@ -0,0 +1,796 @@
...
+				)
+			},
+			expectedErr: prooftypes.ErrProofNotFound.Wrapf(
+				"supplier address %q not found in session ID %q",
+				wrongSupplierAddr,
+				validSessionHeader.GetSessionId(),
+			),
+		},
+		{
+			desc: "merkle proof must be deserializabled",
```
Comment (red-0ne) [inline x/proof/keeper/proof_validation_test.go:270]:
```suggestion
			desc: "merkle proof must be deserializable",
```

File: x/proof/keeper/proof_validation_test.go [hunk 2]
```diff
@@ -0,0 +1,796 @@
...
+			},
+			expectedErr: prooftypes.ErrProofInvalidRelay.Wrapf(
+				"the difficulty relay being proven is (%d), and is smaller than the target difficulty (%d) for service %s",
+				validClosestRelayDifficultyBits,
+				highExpectedTargetDifficulty,
+				validSessionHeader.Service.Id,
+			),
+		},
+		{
+			desc: "relay difficulty must be greater than or equal to minimum (non-zero difficulty)",
```
Comment (red-0ne) [inline x/proof/keeper/proof_validation_test.go:621]: This test is already implemented above.

**Review by red-0ne — CHANGES_REQUESTED** [review]
Tough one, reviewed twice, but not that many change requests. Great job!

File: x/tokenomics/keeper/settle_pending_claims.go [hunk 3]
```diff
@@ -80,14 +80,30 @@ func (k Keeper) SettlePendingClaims(ctx sdk.Context) (
...
-			// If a proof is not found, the claim will expire and never be settled.
-			if !isProofFound {
+		proofIsRequired := (proofRequirement != prooftypes.ProofRequirementReason_NOT_REQUIRED)
+		if proofIsRequired {
+			// EXPIRATION_REASON_UNSPECIFIED is the default
+			var expirationReason types.ClaimExpirationReason = types.ClaimExpirationReason_EXPIRATION_REASON_UNSPECIFIED
+			if isProofFound {
+				var isProofValid bool
+				isProofValid, err = k.proofKeeper.IsProofValid(ctx, &proof)
+				if !isProofValid || err != nil {
```
[the remark this answers is not in the document]
Comment (Olshansk) [reply x/tokenomics/keeper/settle_pending_claims.go:89]: Great suggestion! Done1

File: x/proof/keeper/proof_validation.go [hunk 4]
```diff
@@ -0,0 +1,429 @@
...
+	// Retrieve the supplier's public key.
+	supplierAddr := proof.SupplierAddress
+	supplierPubKey, err := k.accountQuerier.GetPubKeyFromAddress(ctx, supplierAddr)
+	if err != nil {
+		return false, err
+	}
+
+	// Validate the session header.
+	var onChainSession *sessiontypes.Session
+	onChainSession, err = k.queryAndValidateSessionHeader(ctx, proof.SessionHeader, supplierAddr)
```
[the remark this answers is not in the document]
Comment (Olshansk) [reply x/proof/keeper/proof_validation.go:67]: Good point. I'm going to update some comments. PTAL

File: x/proof/keeper/proof_validation.go [hunk 5]
```diff
@@ -0,0 +1,429 @@
...
+	logger.Info("queried and validated the session header")
+
+	// Re-hydrate message session header with the on-chain session header.
+	// This corrects for discrepancies between unvalidated fields in the session
+	// header which can be derived from known values (e.g. session end height).
+	sessionHeader := onChainSession.GetHeader()
+
+	// Validate proof message commit height is within the respective session's
+	// proof submission window using the on-chain session header.
+	if err = k.validateProofWindow(ctx, sessionHeader, supplierAddr); err != nil {
```
[the remark this answers is not in the document]
Comment (Olshansk) [reply x/proof/keeper/proof_validation.go:80]: Good point. I'm going to update some comments. PTAL

**Review by red-0ne — APPROVED** [review]
There is a linter error that is making CI to fail
image

Otherwise LGTM!

**Pull request merged**